### PR TITLE
Automatic Partial Shader Reloading

### PR DESCRIPTION
**Pull request by Tyler-IN**
Makes shader load failures non-fatal.

Carries includes over to to ParsedShader.

### CODE REVIEW

File: Robust.Client/Console/Commands/Debug.cs [hunk 1]
```diff
@@ -735,25 +737,175 @@ public bool Execute(IDebugConsole console, params string[] args)
...
+                {
+                    if (_watchers != null)
+                    {
+                        console.AddLine("Already watching.");
+                        return false;
+                    }
+                    resC = IoCManager.Resolve<IResourceCache>();
+
+                    _watchers = new Dictionary<string, FileSystemWatcher>();
+                    var stringComparer = RuntimeInformation.IsOSPlatform(OSPlatform.Windows) ? StringComparer.OrdinalIgnoreCase : StringComparer.Ordinal;
```
Comment (PJB3005) [inline Robust.Client/Console/Commands/Debug.cs]: macOS is also case insensitive, and Linux can do it too (although you won't see it often). This selection isn't correct.

File: Robust.Client/ResourceManagement/ResourceTypes/ShaderSourceResource.cs [hunk 2]
```diff
@@ -30,10 +32,24 @@ public override void Load(IResourceCache cache, ResourcePath path)
...
+                try
+                {
+                    using var stream = cache.ContentFileRead(path);
+                    using var reader = new StreamReader(stream, EncodingHelpers.UTF8);
+                    ParsedShader = ShaderParser.Parse(reader, cache);
+                    break;
+                }
+                catch (IOException ioe)
+                {
+                    if (!PathHelpers.IsFileInUse(ioe))
```
Comment (PJB3005) [inline Robust.Client/ResourceManagement/ResourceTypes/ShaderSourceResource.cs:48]: Is there no way to use file sharing modes to allow concurrent access?

Comment (PJB3005) [reply Robust.Client/ResourceManagement/ResourceTypes/ShaderSourceResource.cs:48]: Also maybe put a cap on the amount of attempts in this loop.

File: Robust.Client/Graphics/Clyde/Clyde.LightRendering.cs [hunk 3]
```diff
@@ -172,10 +173,19 @@ private void LoadLightingShaders()
...
-                var shaderSource = _resourceCache.GetResource<ShaderSourceResource>(path);
-                return shaderSource.ClydeHandle;
+                try
+                {
+                    var shaderSource = _resourceCache.GetResource<ShaderSourceResource>(path);
+                    return shaderSource.ClydeHandle;
+                }
+                catch (Exception ex)
+                {
+                    Logger.Error($"Can't load shader {path}, {ex.GetType().Name}: {ex.Message}\n{ex.StackTrace}");
```
Comment (PJB3005) [inline Robust.Client/Graphics/Clyde/Clyde.LightRendering.cs]: Can't you achieve all this `{ex.GetType().Name}: {ex.Message}\n{ex.StackTrace}` stuff by just... running `ToString()` on the exception?

**Review by PJB3005 — COMMENTED** [review]
Copying over review comments from the first PR (and added some more):

File: Robust.Client/Console/Commands/Debug.cs (hunk 1 above)
Comment (Tyler-IN) [reply Robust.Client/Console/Commands/Debug.cs]: Got a means to determine case sensitivity?

File: Robust.Client/ResourceManagement/ResourceTypes/ShaderSourceResource.cs (hunk 2 above)
Comment (Tyler-IN) [reply Robust.Client/ResourceManagement/ResourceTypes/ShaderSourceResource.cs:48]: Whoever gets the handle to the file first gets to determine the sharing mode.

File: Robust.Client/Graphics/Clyde/Clyde.LightRendering.cs (hunk 3 above)
Comment (Tyler-IN) [reply Robust.Client/Graphics/Clyde/Clyde.LightRendering.cs]: Yeah. I'll remove ex.StackTrace

File: Robust.Client/Console/Commands/Debug.cs (hunk 1 above)
Comment (PJB3005) [reply Robust.Client/Console/Commands/Debug.cs]: I think on Linux you have to check the directory attributes for a specific flag. On Windows/macOS you can just assume insensitive.

Assuming Linux is case sensitive is fine though. Even CoreCLR does it.

**Pull request merged**